generator/README.md

```diff
@@ -0,0 +1,95 @@
+# Custom Resources Generator
+
+Up until release v0.12 metallb was configured via metallb's Configmap.
+Deprecating the configmap impacts on the users' current configurations.
+To mitigate this, the MetalLB resource generator is an offline tool to
+convert the configmap into the new CRDs.
+
+## Using the generator
+
+### Prerequisites
+
+To use the generator, first create the container image:
+```
+inv build-cr-generator
+```
+### Procedure
+
+Run the container and redirect the output to a file:
+```
+docker run -v <PATH_TO_CONFIG>:/var/input <IMAGE_ID> -origin <CONFIGMAP_FILE_NAME> > metallb_resources.yaml
+```
+
+## Example
+
+For this MetalLB configmap configurations:
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: metallb-system
+  name: config
+data:
+  config: |
+    peers:
+    - my-asn: 64512
+      peer-asn: 64512
+      peer-address: 10.96.0.100
+    address-pools:
+    - name: my-ip-space
+      protocol: bgp
+      addresses:
+      - 198.51.100.0/24
+      bgp-advertisements:
+      - aggregation-length: 32
+        localpref: 100
+        communities:
+        - 64512:1234
+```
+
+The generator will output the following yaml:
+```
+# This was autogenerated by MetalLB's custom resource generator.
+apiVersion: metallb.io/v1beta2
+kind: BGPPeer
+metadata:
+  creationTimestamp: null
+  name: peer1
+  namespace: metallb-system
+spec:
+  holdTime: 1m30s
+  keepaliveTime: 30s
+  myASN: 64512
+  passwordSecret: {}
```

Comment:
same, I assume this shouldn't show up anywhere (it isn't supported in the configmap iirc)

Reply:
We can't avoid having this printed, unless we convert the field to be a pointer (and we don't want to make a pointer out of it for this).
See https://www.sohamkamani.com/golang/omitempty/#values-that-cannot-be-omitted

Reply:
We can't avoid having this printed, unless we convert the field to be a pointer (and we don't want to make a pointer out of it for this).
See https://www.sohamkamani.com/golang/omitempty/#values-that-cannot-be-omitted